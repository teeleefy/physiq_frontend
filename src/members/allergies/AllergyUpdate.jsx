import { Form, FormGroup, Input, Button, Label, Alert } from "reactstrap";
import { useState, useContext, useEffect } from "react";
import {MemberContext} from "../../auth/UserContext.js";
import PhysiqApi from "../../Api.js";
import { useParams, useNavigate } from "react-router-dom";
import Loading from "../../navigation/Loading.jsx";
import '../styles/Allergy.css'
import { toast } from 'react-toastify';


function AllergyUpdate(){
    let navigate = useNavigate();
    let { allergyId } = useParams();
    const { currentMember } = useContext(MemberContext);
    const [isLoading, setIsLoading] = useState(true);
    const [formMessages, setFormMessages] = useState([]);
    const [updateSuccess, setUpdateSuccess] =useState();
    const [formData, setFormData] = useState({name: "", reaction: "", notes: ""});
    

    useEffect(function loadAllergyInfo() {
        async function getCurrentAllergy() {
            if(currentMember && allergyId){
                try{
                    let allergy = await PhysiqApi.getMemberAllergy(currentMember.id, allergyId);
                    if(!allergy.notes){
                        allergy.notes = "";
                      }
                    setFormData({name: allergy.name, reaction: allergy.reaction, notes: allergy.notes})
                }catch (err) {
                console.error("App loadAllergyInfo: problem loading", err);
                }
            }
           setIsLoading(false); 
        }
        getCurrentAllergy();
      }, [allergyId]);



    //SET FORM DATA WITH GOAL INFO 
    
    async function updateAllergy() {
        try {
          let updatedAllergy= await PhysiqApi.updateAllergy(currentMember.id, allergyId, formData);
          return { success: true };
        } catch (errors) {
          console.error("Save failed", errors);
          return { success: false, errors };
        }
      }


    async function handleSubmit(evt){
        evt.preventDefault();
        let result = await updateAllergy();
        
        if(result.success){
            toast.success('Allergy Updated!', {
                position: "top-center",
                autoClose: 2000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                progress: undefined,
                theme: "dark",
                });
           setFormMessages(['Allergy Updated!'])
           setUpdateSuccess(true);
           navigate("..", { relative: "path"})
        }
        else{
            setFormMessages(result.errors);
            setUpdateSuccess(false);
        }
    };

    async function deleteAllergy() {
        try {
          let deletedAllergy= await PhysiqApi.deleteAllergy(currentMember.id, allergyId);
          return { success: true };
        } catch (errors) {
          console.error("Delete failed", errors);
          return { success: false, errors };
        }
      }

    async function handleDelete(evt){
        evt.preventDefault();
        const shouldDelete = confirm("Are you sure you want to delete?")

        if(shouldDelete){
            let result = await deleteAllergy();
            if(result.success){
            // toast.success('Allergy Deleted!')
            toast.success('Allergy Deleted!', {
                position: "top-center",
                autoClose: 2000,
                hideProgressBar: false,
                closeOnClick: true,
                pauseOnHover: true,
                draggable: true,
                progress: undefined,
                theme: "dark",
                });
            setFormMessages(['Allergy Deleted!'])
            setUpdateSuccess(true);
            navigate("..", { relative: "path"})
            }
            else{
                setFormMessages(result.errors);
                setUpdateSuccess(false);
            }
        }
        
    };
  
    /** Update local state w/curr state of input elem */
  
    const handleChange = evt => {
      const { name, value }= evt.target;
      setFormData(fData => ({
        ...fData,
        [name]: value
      }));
    };

    if(isLoading) return <Loading />;

    return(
        <>
            <Form className="Allergy-Form m-4" >
            <h1 className="Allergy-h1">Update Allergy:</h1>
            <p className="text-secondary">Fields marked with <span className="text-danger">*</span> are required.</p>
            <FormGroup>
                <Label className="Allergy-label" for="name" >
                    <b>Allergy Name</b><span className="text-danger">*</span>
                </Label>
                <Input
                id="name"
                name="name"
                maxLength={50}
                value={formData.name}
                type="text"
                onChange={handleChange}
                />
            </FormGroup>
            <FormGroup>
                <Label className="Allergy-label" for="reaction">
                    <b>Reaction</b><span className="text-danger">*</span>
                </Label>
                <Input
                id="reaction"
                name="reaction"
                maxLength={100}
                value={formData.reaction}
                type="text"
                onChange={handleChange}
                />
                </FormGroup>
            <FormGroup>
                <Label className="Allergy-label" for="notes">
                    <b>Notes</b>
                </Label>
                <Input
                id="notes"
                name="notes"
                maxLength={250}
                value={formData.notes}
                type="textarea"
                onChange={handleChange}
                />
            </FormGroup>
            <p className="text-secondary">{formData.notes.length}/250 Characters</p>
                {formMessages.length
                    ? formMessages.map(msg => <Alert color={updateSuccess ? "success": "danger"}>{msg}</Alert>)
                    : null
                }
                <Button className="btn-danger m-2" onClick={handleDelete}>DELETE ALLERGY</Button>
                <Button className="btn-dark m-2" onClick={handleSubmit}>Save Changes</Button>
            </Form>
        </>
    )}
    
export default AllergyUpdate;