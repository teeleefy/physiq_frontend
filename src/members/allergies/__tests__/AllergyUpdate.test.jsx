import {renderWithRouterAndMemberContext, screen} from '../../../../__tests__/test-tools';
import AllergyUpdate from '../AllergyUpdate';

describe('AllergyUpdate', () => {
    it('renders the AllergyUpdate component', () => {
      renderWithRouterAndMemberContext(<AllergyUpdate/>);
      screen.debug(); // prints out the jsx in the App component unto the command line
    });

    it("matches snapshot", function() {
      const { asFragment } = renderWithRouterAndMemberContext(<AllergyUpdate />);
      expect(asFragment()).toMatchSnapshot();
    });

  });
