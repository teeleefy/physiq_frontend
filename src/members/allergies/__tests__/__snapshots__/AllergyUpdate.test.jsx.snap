// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AllergyUpdate > matches snapshot 1`] = `
<DocumentFragment>
  <form
    class="Allergy-Form m-4"
  >
    <h1
      class="Allergy-h1"
    >
      Update Allergy:
    </h1>
    <p
      class="text-secondary"
    >
      Fields marked with 
      <span
        class="text-danger"
      >
        *
      </span>
       are required.
    </p>
    <div
      class="mb-3"
    >
      <label
        class="Allergy-label form-label"
        for="name"
      >
        <b>
          Allergy Name
        </b>
        <span
          class="text-danger"
        >
          *
        </span>
      </label>
      <input
        class="form-control"
        id="name"
        maxlength="50"
        name="name"
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="Allergy-label form-label"
        for="reaction"
      >
        <b>
          Reaction
        </b>
        <span
          class="text-danger"
        >
          *
        </span>
      </label>
      <input
        class="form-control"
        id="reaction"
        maxlength="100"
        name="reaction"
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="Allergy-label form-label"
        for="notes"
      >
        <b>
          Notes
        </b>
      </label>
      <textarea
        class="form-control"
        id="notes"
        maxlength="250"
        name="notes"
      />
    </div>
    <p
      class="text-secondary"
    >
      0/250 Characters
    </p>
    <button
      class="btn-danger m-2 btn btn-secondary"
      type="button"
    >
      DELETE ALLERGY
    </button>
    <button
      class="btn-dark m-2 btn btn-secondary"
      type="button"
    >
      Save Changes
    </button>
  </form>
</DocumentFragment>
`;
